include/unifex/async_destroy.hpp

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License Version 2.0 with LLVM Exceptions
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *   https://llvm.org/LICENSE.txt
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <unifex/just.hpp>
+#include <unifex/tag_invoke.hpp>
+
+namespace unifex {
+namespace _async_destroy {
+
+inline constexpr struct _destroy_fn {
+private:
+  template <typename Resource>
+  static auto try_invoke_member(Resource& resource, int) noexcept
+      -> decltype(resource.destroy()) {
+    return resource.destroy();
+  }
+
+  template <typename Resource>
+  [[deprecated("noop async_destroy")]] static auto
+  try_invoke_member(Resource&, double) noexcept {
+    return unifex::just();
+  }
+
+public:
+  template <typename Resource>
+  constexpr auto operator()(Resource& resource) const noexcept {
+    if constexpr (unifex::tag_invocable<_destroy_fn, Resource&>) {
+      return unifex::tag_invoke(_destroy_fn{}, resource);
+    } else {
+      return try_invoke_member(resource, 1);
+    }
+  }
```

Comment:
If we make `try_invoke_member` conditionally-`noexcept` then this probably needs a rewrite.